File: src/DataScience/Julia/factory_method.jl
struct Database
    connect::Function
    query::Function
end

create_postgres() = Database(
    () -> println("PostgreSQL connect"),
    () -> println("PostgreSQL query"),
)

create_mysql() = Database(
    () -> println("MySQL connect"),
    () -> println("MySQL query"),
)

function use_database(create_database::Function)
    database = create_database()
    database.connect()
    database.query()
end

use_database(create_postgres)
use_database(create_mysql)

File: src/DataScience/R/factory_method.R
create_postgres <- function() {
  list(
    connect = function() cat("PostgreSQL connect\n"),
    query = function() cat("PostgreSQL query\n")
  )
}

create_mysql <- function() {
  list(
    connect = function() cat("MySQL connect\n"),
    query = function() cat("MySQL query\n")
  )
}

use_database <- function(create_database) {
  database <- create_database()
  database$connect()
  database$query()
}

use_database(create_postgres)
use_database(create_mysql)
